In [26]:
import numpy as np
import pandas as pd

#visualization
import matplotlib.pyplot as plt
import seaborn as sns
from PIL import Image as pimg

from IPython.display import display 
pd.options.display.max_columns = None
pd.options.display.max_rows = None

In [6]:
# Train & Test Set Load
train = pd.read_csv("../../csv/train_users_2.csv")
test = pd.read_csv("../../csv/test_users.csv")
ss = pd.read_csv("../../csv/sample_submission_NDF.csv")

In [21]:
# 보조 데이터 Set
age_gender = pd.read_csv("../../csv/age_gender_bkts.csv")
countries = pd.read_csv("../../csv/countries.csv")
sessions = pd.read_csv("../../csv/count_df_agg_sess.csv")

In [9]:
cont_dest = train.filter(['id','country_destination'])

In [10]:
cont_dest

id country_destination
0       gxn3p5htnn                 NDF
1       820tgsjxq7                 NDF
2       4ft3gnwmtx                  US
3       bjjt8pjhuk               other
4       87mebub9p4                  US
5       osr2jwljor                  US
6       lsw9q7uk0j                  US
7       0d01nltbrs                  US
8       a1vcnhxeij                  US
9       6uh8zyj2gn                  US
10      yuuqmid2rp                  US
11      om1ss59ys8                 NDF
12      k6np330cm1                  FR
13      dy3rgx56cu                 NDF
14      ju3h98ch3w                 NDF
15      v4d5rl22px                  CA
16      2dwbwkx056                 NDF
17      frhre329au                  US
18      cxlg85pg1r                 NDF
19      gdka1q5ktd                  FR
20      qdubonn3uk                  US
21      qsibmuz9sx                  US
22      80f7dwscrn                  US
23      jha93x042q                 NDF
24      7i49vnuav6                 NDF
25      al8bcetz0g                  FR
26      bjg0m5otl3                 NDF
27      hfrl5gle36                  US
28      tp6x3md0n4                  FR
29      hql77nu2lk                  US
...            ...                 ...
213421  c98s3h7kgj                 NDF
213422  ytmpiwb8hj                 NDF
213423  3dx1jk6yk2                 NDF
213424  hcfj07iowv                 NDF
213425  l1f71f9vsj                 NDF
213426  15bj4ahmhf                 NDF
213427  qwpybxfjdl                 NDF
213428  k4t61wuvyq                 NDF
213429  mhh7b52z44                 NDF
213430  79wk7k2k5t                 NDF
213431  ftwmocvwlq                 NDF
213432  rg7ayg1tob                 NDF
213433  2f24umzkuv                 NDF
213434  or77n2ojuj                 NDF
213435  0a5bnb9bs4                 NDF
213436  6fzrn49sfn                 NDF
213437  r0jq0devgy                 NDF
213438  v5lq9bj8gv                 NDF
213439  msucfwmlzc                  US
213440  04y8115avm                 NDF
213441  omlc9iku7t                  ES
213442  rf0ay567js                 NDF
213443  0k26r3mir0                  US
213444  40o1ivh6cb                 NDF
213445  qbxza0xojf                  US
213446  zxodksqpep                 NDF
213447  mhewnxesx9                 NDF
213448  6o3arsjbb4                 NDF
213449  jh95kwisub                 NDF
213450  nw9fwlyb5f                 NDF

[213451 rows x 2 columns]

In [22]:
sessions.head()

act_count  C_revert_to_admin  C_events  C_host_standard_suspension  \
0       40.0                0.0       0.0                         0.0   
1       63.0                0.0       0.0                         0.0   
2       90.0                0.0       0.0                         0.0   
3       31.0                0.0       0.0                         0.0   
4      789.0                0.0       0.0                         0.0   

   C_set_minimum_payout_amount  C_set_default  C_rest-of-world  \
0                          0.0            0.0              0.0   
1                          0.0            0.0              0.0   
2                          0.0            0.0              0.0   
3                          0.0            0.0              0.0   
4                          0.0            0.0              0.0   

   C_reset_calendar  C_host_refund_guest  C_report     ...      C_NAN.1  \
0               0.0                  0.0       0.0     ...          3.0   
1               0.0                  0.0       0.0     ...         15.0   
2               0.0                  0.0       0.0     ...          5.0   
3               0.0                  0.0       0.0     ...          0.0   
4               0.0                  0.0       0.0     ...         77.0   

   C_click.1  C_data  C_view  uni_act_type_count  mean_act_type_count  \
0        4.0     9.0    21.0                 7.0                5.715   
1       16.0     9.0    17.0                 6.0               10.500   
2       66.0     2.0     8.0                 6.0               15.000   
3        9.0     5.0    15.0                 5.0                6.200   
4      140.0   140.0   216.0                 8.0               98.600   

   std_act_type_count  med_act_type_count  mode_act_type_count          id  
0               6.777                 3.0                  1.0  00023iyk9l  
1               5.992                12.0                  1.0  0010k6l0om  
2              22.890                 5.5                  2.0  001wyh0pz8  
3               5.305                 5.0                  1.0  0028jgx1x1  
4              77.700               108.5                140.0  002qnbzfs5  

[5 rows x 420 columns]

In [14]:
sessions = sessions.rename(columns={'user_id':'id'})

In [23]:
agg_sees = pd.merge(sessions, cont_dest, on='id', how='left')

In [16]:
agg_sees.age_bkt[(train.age_bkt == '5-9')

id          action action_type        action_detail  \
0  d1mm9tcy42          lookup         NaN                  NaN   
1  d1mm9tcy42  search_results       click  view_search_results   
2  d1mm9tcy42          lookup         NaN                  NaN   
3  d1mm9tcy42  search_results       click  view_search_results   
4  d1mm9tcy42          lookup         NaN                  NaN   

       device_type  secs_elapsed country_destination  
0  Windows Desktop         319.0               other  
1  Windows Desktop       67753.0               other  
2  Windows Desktop         301.0               other  
3  Windows Desktop       22141.0               other  
4  Windows Desktop         435.0               other

In [27]:
data1 = agg_sees.pivot_table(columns=['country_destination'], index=['C_view_search_results'], values='id', aggfunc="count")
display(data1)

country_destination      AU     CA    DE     ES     FR     GB     IT      NDF  \
C_view_search_results                                                           
0.0                    61.0  137.0  75.0  206.0  404.0  224.0  250.0  15241.0   
1.0                     7.0   19.0  10.0   45.0   78.0   39.0   53.0   2952.0   
2.0                     5.0   25.0  14.0   34.0   63.0   27.0   41.0   2638.0   
3.0                     8.0   20.0  14.0   33.0   58.0   29.0   26.0   2102.0   
4.0                     2.0   17.0   8.0   13.0   46.0   31.0   34.0   1856.0   
5.0                     8.0   10.0   6.0   20.0   37.0   22.0   28.0   1570.0   
6.0                     3.0   12.0   6.0   15.0   34.0   15.0   38.0   1394.0   
7.0                     1.0   11.0   2.0   15.0   38.0   20.0   29.0   1275.0   
8.0                     3.0   13.0   7.0   10.0   33.0   16.0   22.0   1082.0   
9.0                     2.0    9.0   5.0   14.0   35.0   14.0   20.0    964.0   
10.0                    2.0    7.0   7.0   18.0   22.0   17.0   20.0    845.0   
11.0                    1.0   12.0   9.0   12.0   27.0   16.0   14.0    781.0   
12.0                    1.0    5.0   3.0   16.0   20.0    8.0   20.0    706.0   
13.0                    NaN    7.0   5.0   13.0   22.0   12.0   12.0    643.0   
14.0                    NaN    6.0   6.0    9.0   27.0   15.0   18.0    606.0   
15.0                    3.0    5.0   4.0    9.0   24.0   10.0    9.0    547.0   
16.0                    2.0    5.0   NaN    6.0   17.0    7.0    8.0    516.0   
17.0                    NaN    5.0   2.0    4.0   14.0    9.0   15.0    488.0   
18.0                    2.0    6.0   NaN   17.0   26.0    8.0   12.0    438.0   
19.0                    NaN    8.0   8.0   10.0   14.0   12.0   10.0    387.0   
20.0                    2.0    7.0   1.0   11.0   19.0   12.0   13.0    392.0   
21.0                    1.0    8.0   2.0    5.0   10.0   10.0   12.0    348.0   
22.0                    2.0   10.0   2.0    9.0   11.0    6.0   12.0    340.0   
23.0                    NaN    3.0   4.0    7.0   10.0    6.0   12.0    308.0   
24.0                    1.0    3.0   1.0    7.0    9.0    5.0    9.0    299.0   
25.0                    2.0    2.0   3.0    5.0    9.0    6.0    7.0    281.0   
26.0                    1.0    2.0   3.0    5.0   13.0    4.0    7.0    236.0   
27.0                    NaN    5.0   3.0    6.0    6.0    8.0    8.0    236.0   
28.0                    NaN    1.0   NaN    4.0   13.0    7.0    5.0    223.0   
29.0                    1.0    2.0   NaN    7.0   15.0    5.0    7.0    218.0   
30.0                    1.0    3.0   NaN    7.0    7.0    2.0    8.0    205.0   
31.0                    3.0    2.0   2.0    2.0    7.0    1.0    9.0    222.0   
32.0                    NaN    1.0   1.0    2.0   13.0    4.0    6.0    216.0   
33.0                    2.0    1.0   1.0    1.0    8.0    4.0    5.0    181.0   
34.0                    1.0    1.0   1.0    3.0   10.0    5.0    8.0    160.0   
35.0                    NaN    1.0   2.0    3.0    9.0    4.0    5.0    156.0   
36.0                    1.0    1.0   NaN    1.0   11.0    3.0    6.0    153.0   
37.0                    2.0    1.0   NaN    1.0    5.0    2.0    3.0    130.0   
38.0                    1.0    3.0   2.0    4.0    4.0    6.0    4.0    116.0   
39.0                    1.0    1.0   2.0    5.0    7.0    1.0    3.0    120.0   
40.0                    2.0    1.0   NaN    2.0    5.0    4.0    5.0    112.0   
41.0                    NaN    1.0   1.0    3.0    6.0    1.0    7.0    127.0   
42.0                    2.0    2.0   NaN    2.0    5.0    NaN    4.0    119.0   
43.0                    NaN    2.0   2.0    4.0    5.0    3.0    3.0    113.0   
44.0                    1.0    1.0   2.0    3.0    7.0    4.0    2.0    105.0   
45.0                    NaN    NaN   NaN    2.0    7.0    1.0    5.0    110.0   
46.0                    1.0    1.0   NaN    2.0    6.0    1.0    3.0     90.0   
47.0                    NaN    

In [ ]:
train.shape, test.shape

In [ ]:
test.isnull().sum()

In [ ]:
test.tail()

In [ ]:
train.info()

In [ ]:
train.isnull().sum()

In [ ]:
for col in train.columns:
    print("{col} : {count}".format(col=col, count=len(train[col].unique())))

In [ ]:
train.tail()

## 1. Target (Destination) 탐색

In [ ]:
print(train['country_destination'].unique())

In [ ]:
# 나라별로 Count 
data1 = train.groupby('country_destination').aggregate({'id':'count'}).reset_index()
data1.sort_values(by=['id'], ascending=False)

Countries Category가 12개

예외 1) NDF : Do Not Make a booking (58.35%)
 
예외 2) others : 4.7%

In [ ]:
124543/213451, 10094/213451

### 1.1 여행지 비율

In [ ]:
plt.figure(figsize=(10,8))
destination_percentage = train.country_destination.value_counts() / train.shape[0] * 100
destination_percentage.plot(kind='bar',color='#FD5C64', rot=0)

plt.ylabel("Loss in ${0}'s".format(value_increment))
plt.yticks(values * value_increment, ['%d' % val for val in values])
plt.xticks([])
plt.title('Loss by Disaster')

plt.show()


plt.xlabel('Destination Country')
plt.ylabel('Percentage')
plt.show()

In [ ]:
destination_percentage.values

## 2. Gender Column 탐색

In [ ]:
data2 = train.groupby('gender').aggregate({'id':'count'}).reset_index()
data2.sort_values(by=['id'], ascending=False)

unknown 값을 어떻게 처리할까?

idea1) train data의 나이를 참조해서 age_gender의 나이대에 성별을 랜덤하게 입력?

idea2) 성별이 영향이 있을까? 동반자 데이터가 없는데. 혼자가 아닌 이상 영향이 없을 것 같다는 생각도 듬 : age_gender 데이터를 통해 극명하게 성별에 따른 우위가 있는 여행지가 있는지 확인

In [ ]:
sns.countplot(x=train.gender, palette='Paired')
plt.show()

### 2.1 성별에 따른 여행지 선호도가 달라질까?

In [ ]:
female_destinations = train.loc[train['gender'] == 'FEMALE', 'country_destination'].value_counts() / sum(train['gender'] == 'FEMALE') * 100
male_destinations = train.loc[train['gender'] == 'MALE', 'country_destination'].value_counts() / sum(train['gender'] == 'MALE') * 100
unknown_destinations = train.loc[train['gender'] == '-unknown-', 'country_destination'].value_counts() / sum(train['gender'] == '-unknown-') * 100

plt.figure(figsize=(10,8))
male_destinations.plot(kind='bar', width=0.2, color='lavender', position=0, label='Male', rot=0)
female_destinations.plot(kind='bar', width=0.2, color='khaki', position=1, label='Female', rot=0)
unknown_destinations.plot(kind='bar', width=0.2, color='orchid', position=2, label='unknown', rot=0)

plt.legend()
plt.xlabel('Destination Country')
plt.ylabel('Percentage')


plt.show()

프랑스, 이탈리아, 영국: 여성 선호 (패션, 아름다움)

캐나다, 독일, 네덜란드 : 남성선호 (Extrme, 맥주, 성, 축구)

## 3. Age Column 탐색 (# of NaN = 87,990)

In [ ]:
plt.figure(figsize=(10,8))

sns.countplot(x=train.age.dropna())
plt.show()

In [5]:
indices = list(train.age[train.age == 2014].index)
for index in indices:
    train.age.loc[index] = np.nan
    
indices = list(train.age[train.age == 2013].index)
for index in indices:
    train.age.loc[index] = np.nan
    
indices = list(train.age[(train.age > 120) & (train.age < 200)].index)
for index in indices:
    train.age.loc[index] = np.nan

/usr/local/lib/python3.6/site-packages/pandas/core/indexing.py:194: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


In [6]:
train = train.copy()
indices = list(train.age[train.age > 120].index)
for index in indices:
    if train.age.loc[index] > 1900:
        mod = 2015 - train.age.loc[index]
        train.age.loc[index] = mod
train.tail()

/usr/local/lib/python3.6/site-packages/pandas/core/indexing.py:194: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


id date_account_created  timestamp_first_active  \
213446  zxodksqpep           2014-06-30          20140630235636   
213447  mhewnxesx9           2014-06-30          20140630235719   
213448  6o3arsjbb4           2014-06-30          20140630235754   
213449  jh95kwisub           2014-06-30          20140630235822   
213450  nw9fwlyb5f           2014-06-30          20140630235824   

       date_first_booking     gender   age signup_method  signup_flow  \
213446                NaN       MALE  32.0         basic            0   
213447                NaN  -unknown-   NaN         basic            0   
213448                NaN  -unknown-  32.0         basic            0   
213449                NaN  -unknown-   NaN         basic           25   
213450                NaN  -unknown-   NaN         basic           25   

       language affiliate_channel affiliate_provider first_affiliate_tracked  \
213446       en         sem-brand             google                     omg   
213447       en            direct             direct                  linked   
213448       en            direct             direct               untracked   
213449       en             other              other           tracked-other   
213450       en            direct             direct               untracked   

       signup_app first_device_type  first_browser country_destination  
213446        Web       Mac Desktop         Safari                 NDF  
213447        Web   Windows Desktop         Chrome                 NDF  
213448        Web       Mac Desktop        Firefox                 NDF  
213449        iOS            iPhone  Mobile Safari                 NDF  
213450        iOS            iPhone      -unknown-                 NDF

In [7]:
df = pd.DataFrame(columns = ["age", "age_bucket"])
df["age"] = list(range(0, 151))

age_bucket = list(age_gender.age_bucket.unique())
age_bucket.reverse()

for i in range(len(df)):
    if df.age[i] > 100:
        df.age_bucket.loc[i] = age_bucket[-1]
    else:
        df.age_bucket.loc[i] = age_bucket[i//5]

df = df.set_index("age")
df

/usr/local/lib/python3.6/site-packages/pandas/core/indexing.py:194: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


age_bucket
age           
0          0-4
1          0-4
2          0-4
3          0-4
4          0-4
5          5-9
6          5-9
7          5-9
8          5-9
9          5-9
10       10-14
11       10-14
12       10-14
13       10-14
14       10-14
15       15-19
16       15-19
17       15-19
18       15-19
19       15-19
20       20-24
21       20-24
22       20-24
23       20-24
24       20-24
25       25-29
26       25-29
27       25-29
28       25-29
29       25-29
..         ...
121       100+
122       100+
123       100+
124       100+
125       100+
126       100+
127       100+
128       100+
129       100+
130       100+
131       100+
132       100+
133       100+
134       100+
135       100+
136       100+
137       100+
138       100+
139       100+
140       100+
141       100+
142       100+
143       100+
144       100+
145       100+
146       100+
147       100+
148       100+
149       100+
150       100+

[151 rows x 1 columns]

In [8]:
train["age_bkt"] = train["age"].map(df.age_bucket)
train.head()

id date_account_created  timestamp_first_active date_first_booking  \
0  gxn3p5htnn           2010-06-28          20090319043255                NaN   
1  820tgsjxq7           2011-05-25          20090523174809                NaN   
2  4ft3gnwmtx           2010-09-28          20090609231247         2010-08-02   
3  bjjt8pjhuk           2011-12-05          20091031060129         2012-09-08   
4  87mebub9p4           2010-09-14          20091208061105         2010-02-18   

      gender   age signup_method  signup_flow language affiliate_channel  \
0  -unknown-   NaN      facebook            0       en            direct   
1       MALE  38.0      facebook            0       en               seo   
2     FEMALE  56.0         basic            3       en            direct   
3     FEMALE  42.0      facebook            0       en            direct   
4  -unknown-  41.0         basic            0       en            direct   

  affiliate_provider first_affiliate_tracked signup_app first_device_type  \
0             direct               untracked        Web       Mac Desktop   
1             google               untracked        Web       Mac Desktop   
2             direct               untracked        Web   Windows Desktop   
3             direct               untracked        Web       Mac Desktop   
4             direct               untracked        Web       Mac Desktop   

  first_browser country_destination age_bkt  
0        Chrome                 NDF     NaN  
1        Chrome                 NDF   35-39  
2            IE                  US   55-59  
3       Firefox               other   40-44  
4        Chrome                  US   40-44

In [46]:
indices = list(train.age_bkt[(train.age_bkt == '0-4')].index)

for index in indices:
    train.age_bkt.loc[index] = 'minor'

In [49]:
indices = list(train.age_bkt[(train.age_bkt == '5-9')].index)

for index in indices:
    train.age_bkt.loc[index] = 'minor'

/usr/local/lib/python3.6/site-packages/pandas/core/indexing.py:194: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


In [50]:
indices = list(train.age_bkt[(train.age_bkt == '10-14')].index)

for index in indices:
    train.age_bkt.loc[index] = 'minor'

In [51]:
indices = list(train.age_bkt[(train.age_bkt == '15-19')].index)

for index in indices:
    train.age_bkt.loc[index] = 'minor'

/usr/local/lib/python3.6/site-packages/pandas/core/indexing.py:194: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


In [54]:
indices = list(train.age[(train.age_bkt == '70-74') | (train.age_bkt == '75-79') |
                        (train.age_bkt == '80-84')| (train.age_bkt == '85-89')|
                        (train.age_bkt == '90-94')| (train.age_bkt == '95-99')].index)
for index in indices:
    train.age_bkt.loc[index] = 'senior'

/usr/local/lib/python3.6/site-packages/pandas/core/indexing.py:194: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


In [55]:
train[train.age == 76]

id date_account_created  timestamp_first_active  \
1822    souftdfcq1           2010-09-21          20100921012022   
12690   f5mkm7otyx           2011-11-16          20111116182318   
12991   cd6nfc2ojg           2011-11-23          20111123112735   
17577   vh9f26dav4           2012-02-24          20120224005730   
22787   cudfyjx9d0           2012-05-01          20120501035448   
32819   x71ehgib7x           2012-07-19          20120719235220   
33783   y59lhiaxbk           2012-07-26          20120726183837   
33787   zll4nebvv9           2012-07-26          20120726192148   
35968   3iw3zvape4           2012-08-10          20120810224304   
43185   x7x7dsdhtt           2012-10-01          20121001231920   
43952   q3undyj57v           2012-10-08          20121008014541   
46087   fn6sx322wx           2012-10-25          20121025010605   
46211   rx9h9lzbe0           2012-10-26          20121026020036   
47241   vgkiocestb           2012-11-04          20121104023807   
49529   4v3dfnsnx4           2012-11-23          20121123023111   
53456   jozjdnvn4c           2012-12-26          20121226171800   
53942   emd6wcxmra           2012-12-30          20121230191335   
54519   w1oy7thbsw           2013-01-04          20130104181754   
55617   4avuj0qpcp           2013-01-12          20130112014719   
58299   22lopo0z9l           2013-01-30          20130130210857   
59031   pteuqip2h5           2013-02-05          20130205044954   
59818   bncqlulxtn           2013-02-10          20130210022440   
63511   3ybjqy07tf           2013-03-04          20130304234140   
64051   z9c8qrgp4e           2013-03-07          20130307215115   
64514   ukgekpm12f           2013-03-10          20130310230325   
65430   0vyj7ieft4           2013-03-15          20130315151149   
67446   2pxaa0yk38           2013-03-26          20130326233718   
69046   3zztgmrvp4           2013-04-04          20130404162503   
72064   eluucuraw6           2013-04-20          20130420185538   
73508   cd11tb3scl           2013-04-28          20130428012327   
...            ...                  ...                     ...   
101486  xabxmctfg4           2013-08-22          20130822160502   
112107  fnjr3yu1p6           2013-09-28          20130928004918   
114862  tptl6m1hjy           2013-10-07          20131007065159   
118388  60p95njq9n           2013-10-21          20131021062416   
124912  ce87xhk3ne           2013-11-16          20131116010813   
124926  ony0mr4t8c           2013-11-16          20131116015823   
125738  ur104ynzxy           2013-11-19          20131119021428   
126533  iaze0jbak8           2013-11-21          20131121175606   
137033  ucgks2fyez           2014-01-01          20140101010900   
145929  q1vzpf4vp1           2014-01-25          20140125215252   
148735  lcyq1gf39q           2014-02-03          20140203040302   
153814  tdp37lr7lp           2014-02-17          20140217150929   
156319  8sa3abisbl           2014-02-24          20140224160149   
157939  pcsivaznbu           2014-02-28          20140228161018   
158043  4b6sxjqzau           2014-02-28          20140228222831   
159730  h5h76u2iht           2014-03-05          20140305174147   
160526  gtq2sws3xt           2014-03-07          20140307164407   
167190  s2ifxbv2ib           2014-03-24          20140324210551   
168429  g1hrbapb04           2014-03-27          20140327153755   
170122  9m0iyjoxod           2014-03-31          20140331232659   
172511  ztja1nhito           2014-04-07          20140407002129   
178593  z6ghj2bx1k           2014-04-22          20140422000308   
180220  kv2veaehuq           2014-04-25          20140425022650   
183807  sveftds5jz           2014-05-03          20140503014553   
184877  lg5xi23ryr           2014-05-05          20140505190938   
189012  3myzvwd9eu           2014-05-14          20140514051044   
189460  cnlqs7a9rg           2014-05-15          20140515002932   
189493  cwjifq8oeq           2014-05-15          20140515012806

## Feature 정리 (100 이하 others)

In [ ]:
train.affiliate_provider.value_counts()

In [9]:
list_afiliate = train.affiliate_provider.value_counts()
df_afiliate = pd.DataFrame(list_afiliate)
afiliate_list = df_afiliate[df_afiliate['affiliate_provider']>100].index
afiliate_list
train['affiliate_provider'] = train['affiliate_provider'].apply(lambda x : x if x in afiliate_list else 'others')

In [ ]:
train.affiliate_provider.value_counts()

In [ ]:
train.first_device_type.value_counts()

In [ ]:
~19세 / 70세~

### Age Modeling

In [10]:
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.cross_validation import cross_val_score

/usr/local/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [56]:
age_train = train[train.age_bkt.notnull()]
age_train

id date_account_created  timestamp_first_active  \
1       820tgsjxq7           2011-05-25          20090523174809   
2       4ft3gnwmtx           2010-09-28          20090609231247   
3       bjjt8pjhuk           2011-12-05          20091031060129   
4       87mebub9p4           2010-09-14          20091208061105   
6       lsw9q7uk0j           2010-01-02          20100102012558   
7       0d01nltbrs           2010-01-03          20100103191905   
8       a1vcnhxeij           2010-01-04          20100104004211   
9       6uh8zyj2gn           2010-01-04          20100104023758   
10      yuuqmid2rp           2010-01-04          20100104194251   
11      om1ss59ys8           2010-01-05          20100105051812   
13      dy3rgx56cu           2010-01-05          20100105083259   
14      ju3h98ch3w           2010-01-07          20100107055820   
15      v4d5rl22px           2010-01-07          20100107204555   
17      frhre329au           2010-01-07          20100107224625   
19      gdka1q5ktd           2010-01-10          20100110010817   
21      qsibmuz9sx           2010-01-10          20100110220941   
22      80f7dwscrn           2010-01-11          20100111031438   
24      7i49vnuav6           2010-01-11          20100111230808   
25      al8bcetz0g           2010-01-12          20100112131444   
27      hfrl5gle36           2010-01-12          20100112205949   
28      tp6x3md0n4           2010-01-13          20100113044650   
29      hql77nu2lk           2010-01-13          20100113064333   
30      cheova4spt           2010-01-14          20100114025257   
31      pggg2sj27u           2010-01-14          20100114060757   
32      mo2v5h3nti           2010-01-14          20100114084926   
33      qthj88nnc7           2010-01-14          20100114175209   
34      x9gnxoun57           2010-01-15          20100115055053   
35      ugy4obax11           2010-01-15          20100115082711   
36      7my0vrljxc           2010-01-15          20100115221929   
37      ul0b29nhr6           2010-01-15          20100115231651   
...            ...                  ...                     ...   
213390  c999p94qui           2014-06-30          20140630222253   
213391  nb431o6mov           2014-06-30          20140630222254   
213393  2ntc2mstbz           2014-06-30          20140630223010   
213395  otum1j4w9u           2014-06-30          20140630223150   
213398  lc6apdi0hc           2014-06-30          20140630223731   
213400  x17avasiuk           2014-06-30          20140630223943   
213401  xif0n17eof           2014-06-30          20140630224049   
213402  jfcwjhkmtv           2014-06-30          20140630224108   
213403  f9bs0rczyv           2014-06-30          20140630224150   
213405  5q5k8aov71           2014-06-30          20140630224452   
213406  x3m80xzv0h           2014-06-30          20140630224540   
213407  iu1abga5nx           2014-06-30          20140630224635   
213408  l8tltghomx           2014-06-30          20140630224654   
213409  git9gx54o9           2014-06-30          20140630224802   
213410  ezuin5g4lu           2014-06-30          20140630224832   
213412  a1rvlsdijt           2014-06-30          20140630225157   
213415  7plqriggxp           2014-06-30          20140630230136   
213417  zhf38bsj36           2014-06-30          20140630230934   
213423  3dx1jk6yk2           2014-06-30          20140630231548   
213424  hcfj07iowv           2014-06-30          20140630231859   
213425  l1f71f9vsj           2014-06-30          20140630232119   
213430  79wk7k2k5t           2014-06-30          20140630233132   
213432  rg7ayg1tob           2014-06-30          20140630233224   
213439  msucfwmlzc           2014-06-30          20140630234729   
213440  04y8115avm           2014-06-30          20140630234933   
213441  omlc9iku7t           2014-06-30          20140630235151   
213443  0k26r3mir0           2014-06-30          20140630235340   
213445  qbxza0xojf           2014-06-30          20140630235547

In [57]:
age_test = train[train.age_bkt.isnull()]
age_test

id date_account_created  timestamp_first_active  \
0       gxn3p5htnn           2010-06-28          20090319043255   
5       osr2jwljor           2010-01-01          20100101215619   
12      k6np330cm1           2010-01-05          20100105060859   
16      2dwbwkx056           2010-01-07          20100107215125   
18      cxlg85pg1r           2010-01-08          20100108015641   
20      qdubonn3uk           2010-01-10          20100110152120   
23      jha93x042q           2010-01-11          20100111224015   
26      bjg0m5otl3           2010-01-12          20100112155420   
38      7yzz3iv0ji           2010-01-15          20100115234132   
43      69xzq1jxkq           2010-01-21          20100121183725   
44      g3z67e37mb           2010-12-01          20100123090523   
46      f6mzhb4wbc           2010-01-24          20100124002335   
47      ngai8kt2yi           2010-01-24          20100124020049   
48      2cji4lpjr3           2010-01-24          20100124083507   
51      rcx09dn2je           2010-01-25          20100125203143   
57      ehcdcpkn8y           2010-01-27          20100127083836   
59      rwa710uoxj           2010-01-28          20100128100127   
60      p968flw686           2010-01-29          20100129032157   
62      5dp8dxiayg           2010-01-29          20100129173619   
63      ranh94tmin           2010-01-29          20100129182105   
66      lfwwfmbi0g           2010-01-31          20100131003213   
67      mz47bohqc8           2010-02-01          20100201021621   
68      nvti2htxna           2010-02-01          20100201061326   
69      o47imoy58k           2010-02-01          20100201224759   
70      b9gu8uoxxw           2010-02-02          20100202050231   
73      61rgez44q4           2010-02-02          20100202180421   
78      jf5oxeb233           2010-02-04          20100204180032   
80      jzw5p68a5u           2010-02-06          20100206070403   
82      odtqrxp6d1           2010-02-06          20100206155410   
83      yzjoqp3401           2010-02-06          20100206163608   
...            ...                  ...                     ...   
213394  kxbz3jr8je           2014-06-30          20140630223056   
213396  5vtfm5aila           2014-06-30          20140630223155   
213397  hprdog4tf0           2014-06-30          20140630223618   
213399  9r0b1sd15r           2014-06-30          20140630223741   
213404  e5ysdcsb2v           2014-06-30          20140630224336   
213411  16d2b7lqz2           2014-06-30          20140630225156   
213413  qutd4hqiss           2014-06-30          20140630225543   
213414  hlelejm47r           2014-06-30          20140630225637   
213416  pa8rnz3ne0           2014-06-30          20140630230253   
213418  u6h23bsnzv           2014-06-30          20140630230950   
213419  whl3u790pe           2014-06-30          20140630231044   
213420  so90y4lyii           2014-06-30          20140630231101   
213421  c98s3h7kgj           2014-06-30          20140630231137   
213422  ytmpiwb8hj           2014-06-30          20140630231246   
213426  15bj4ahmhf           2014-06-30          20140630232331   
213427  qwpybxfjdl           2014-06-30          20140630232539   
213428  k4t61wuvyq           2014-06-30          20140630232634   
213429  mhh7b52z44           2014-06-30          20140630232712   
213431  ftwmocvwlq           2014-06-30          20140630233203   
213433  2f24umzkuv           2014-06-30          20140630233427   
213434  or77n2ojuj           2014-06-30          20140630233640   
213435  0a5bnb9bs4           2014-06-30          20140630233851   
213436  6fzrn49sfn           2014-06-30          20140630234113   
213437  r0jq0devgy           2014-06-30          20140630234243   
213438  v5lq9bj8gv           2014-06-30          20140630234429   
213442  rf0ay567js           2014-06-30          20140630235309   
213444  40o1ivh6cb           2014-06-30          20140630235352   
213447  mhewnxesx9           2014-06-30          20140630235719

In [58]:
feature_names = ['signup_method', 'signup_app', 'affiliate_channel',
                 'affiliate_provider', 'first_device_type', 'first_browser']
dfX = age_train[feature_names].copy()
dfy = age_train['age_bkt'].copy()

In [ ]:
dfX.tail()

In [59]:
dummies = ['signup_method', 'signup_app', 'affiliate_channel',
          'affiliate_provider', 'first_device_type', 'first_browser']
for feature in dummies:
    dummy_features = pd.get_dummies(dfX[feature], prefix=feature)
    for dummy in dummy_features:
        dfX[dummy] = dummy_features[dummy]
    dfX = dfX.drop([feature], 1)

## Feature selection 2

In [ ]:
feature_names = ['signup_method', 'signup_app', 'affiliate_channel',
                 'affiliate_provider', 'first_device_type', 'first_browser']
dfX = age_train[feature_names].copy()
dfy = age_train['age_bkt'].copy()

In [62]:

dfy2 = pd.DataFrame(LabelEncoder().fit_transform(dfy), index=dfy.index)
dfy2

0
1        4
2        8
3        5
4        5
6        6
7        6
8        7
9        6
10       4
11       6
13       4
14       4
15       3
17       3
19       2
21       3
22       5
24       5
25       2
27       3
28       4
29       4
30       5
31       3
32       3
33       2
34       8
35       6
36       3
37       3
...     ..
213390   3
213391   2
213393   2
213395   3
213398   4
213400   7
213401   6
213402   3
213403   2
213405   1
213406   4
213407   1
213408  10
213409   2
213410   3
213412   7
213415   8
213417   6
213423   1
213424   3
213425   3
213430  11
213432   3
213439   5
213440   1
213441   3
213443   4
213445   1
213446   3
213448   3

[124710 rows x 1 columns]

In [ ]:
# from sklearn.preprocessing import LabelBinarizer
# dfy3 = pd.DataFrame(LabelBinarizer().fit_transform(dfy), index=dfy.index)
# dfy3

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(dfX, dfy2, test_size=0.25, random_state=0)

In [63]:
from sklearn.tree import DecisionTreeClassifier
model = DecisionTreeClassifier(criterion='entropy', max_depth=10, min_samples_leaf=1).fit(dfX, dfy2)

In [64]:
from sklearn.metrics import confusion_matrix

result = confusion_matrix(dfy2, model.predict(dfX))

In [ ]:
X_train.columns[35]

In [ ]:
command_buf = io.StringIO() 
export_graphviz(model, out_file=command_buf) 
graph = pydot.graph_from_dot_data(command_buf.getvalue())[0] 
image = graph.create_png()
#Image(image)
# newImg1 = pimg.new('RGB', (512,512))
# newImg1.pimg.save("img1.png")

In [22]:
pd.DataFrame(data=result[1:,1:], index=result[1:,0], columns=result[0,1:])

0  0   0      4      8    0   0   0  0   0   0   0  0  0  0  0  0  0  0
0  3  0   3    438   1111   29   1   1  0   2   0   1  1  0  0  0  0  0  0
0  0  3  22    533   1281   27   0   2  0   3   1   0  0  0  0  0  0  0  0
0  0  0  98   4877   3870   48   4   3  0   6   0   0  0  1  0  0  0  0  0
0  0  1  57  13149  13755  145  10   7  0  16   1   1  0  1  0  0  0  0  0
0  2  2  66  11287  16915  222  11   8  0  27   5   4  1  1  0  0  0  0  0
0  1  1  54   6722  11898  286   9  13  0  26   1   8  0  0  0  0  0  0  0
0  0  0  54   3694   7736  165  34  16  0  25   5   9  1  1  0  0  0  0  0
0  0  0  39   2322   5856  154   8  41  0  30  13   5  2  0  0  0  0  0  0
0  0  0   0     15     30    0   0   0  0   1   0   0  0  0  0  0  0  0  0
0  0  1  31   1444   4338  141  10  19  0  51   5   9  2  0  0  0  0  0  0
0  0  0  24    934   3306  120   8  19  0  30  22   7  0  0  0  0  0  0  0
0  0  1  15    590   2372   82   3  18  0  23   8  18  1  0  0  0  0  0  0
0  1  0   8    349   1581   48   0   4  0  10   1   2  3  0  0  0  0  0  0
0  0  0   3    176    672   29   0   3  0  10   4   0  2  2  0  0  0  0  0
0  0  0   2     65    259    4   0   2  0   0   2   0  3  0  0  0  0  0  0
0  0  0   1     43    106    3   0   1  0   1   0   0  0  0  0  0  0  0  0
0  0  0   1     41     69    1   1   0  0   1   0   0  1  0  0  0  0  0  0
0  0  0   1     33     41    1   0   0  0   1   0   0  0  0  0  0  0  0  0
0  0  0   0     48     65    1   0   1  0   1   0   0  0  0  0  0  0  0  0

In [65]:
from sklearn.metrics import classification_report

print(classification_report(dfy2, model.predict(dfX)))

             precision    recall  f1-score   support

          0       0.50      0.00      0.00      1590
          1       0.21      0.01      0.02      8907
          2       0.28      0.48      0.36     27143
          3       0.22      0.59      0.33     28551
          4       0.19      0.02      0.03     19019
          5       0.34      0.00      0.01     11740
          6       0.26      0.00      0.01      8470
          7       0.18      0.01      0.01      6051
          8       0.35      0.01      0.01      4470
          9       0.29      0.01      0.01      3131
         10       0.00      0.00      0.00      2007
         11       0.00      0.00      0.00      1930
         12       0.35      0.00      0.01      1701

avg / total       0.25      0.25      0.16    124710



/usr/local/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [ ]:
n_estimators = 100
model_rf = RandomForestClassifier(max_depth=4, n_estimators=n_estimators).fit(X_train, y_train)

In [ ]:
print(classification_report(y_train, model_rf.predict(X_train)))

In [ ]:
train2 = train

bins = [0, 30, 60, 100]
group_names = ["younger", "family", "older"]
train2['group_age'] = pd.cut(train['age'], bins, labels=group_names, right=False)

In [ ]:
train2['group_age'].dropna(axis=0, how='any')

In [ ]:
younger = sum(train2['group_age'] == 'younger')
family = sum(train2['group_age'] == 'family')
older = sum(train2['group_age'] == 'older')

younger_destinations = train2.loc[train2['group_age'] == 'younger', 'country_destination'].value_counts() / younger * 100
family_destinations = train2.loc[train2['group_age'] == 'family', 'country_destination'].value_counts() / family * 100
older_destinations = train2.loc[train2['group_age'] == 'older', 'country_destination'].value_counts() / older * 100

print("young", "\n", younger_destinations,"\n", "family", "\n", family_destinations,"\n", "older", "\n", older_destinations)

plt.figure(figsize=(10,8))
younger_destinations.plot(kind='bar', width=0.2, color='r', position=0, label='Youngers', rot=0)
family_destinations.plot(kind='bar', width=0.2, color='b', position=1, label='Family', rot=0)
older_destinations.plot(kind='bar', width=0.2, color='g', position=2, label='Olders', rot=0)


plt.legend()
plt.xlabel('Destination Country')
plt.ylabel('Percentage')

plt.show()


In [ ]:
train2 = train

bins = [0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
group_names = ["u10", "10g", "20g", "30g", "40g", "50g", "60g", "70g",
              "80g", "90g"]
train2['group_gen'] = pd.cut(train2['age'], bins, labels=group_names, right=False)
train2['group_gen'].dropna(axis=0, how='any')

### 세대별 여행지 선택

In [ ]:
data3_2 = train2.groupby('group_gen').aggregate({'id':'count'}).reset_index()
data3_2

In [ ]:
data3_2 = train2.pivot_table(columns='group_gen', index=['country_destination'],values='id', aggfunc="count")
data3_2.fillna(0)

In [ ]:
plt.figure(figsize=(10,8))
labels = 'AU', 'CA', 'DE', 'ES', 'FR', 'GB', 'IT', 'NDF', 'NL', 'PT', 'US', 'other'
sizes = data3_2['20g'].values.tolist()
plt.pie(sizes, autopct='%1.1f%%', labels=labels, shadow=True, startangle=140)
plt.tight_layout()
plt.show()

In [ ]:
age_gender[age_gender['age_bucket']  == '20-24']

In [ ]:
train

In [ ]:
dfx = train2.drop(['age_bkt', 'id','date_account_created','timestamp_first_active', 'date_first_booking', 'gender','language', 'first_affiliate_tracked', 'country_destination', 'signup_flow', 'age'],1)

In [ ]:
dfx.head()

In [ ]:
y = train2['age_bkt']

In [ ]:
from sklearn.tree import DecisionTreeRegressor

In [ ]:
data3 = train[train.age > 100]
len(data3)

In [ ]:
data3_1 = data3.groupby('age').aggregate({'id':'count'}).reset_index()
data3_1

기네스북 최고령자는 146세, 가장 최근은 116세

1924~2008은 생년월일을 잘못 입력한 듯

2013, 2014는 해당년도로 잘못분류 된듯

나이 데이터는 중요할 것으로 판단되는데... 어떻게 처리할지...

In [ ]:
age_gender.tail()

또 다른 조사기관인 닐슨온라인(Nielsen Online)의 조사에서도 이와 유사한 경향이 나타났음. 젊은 계층에 비해 나이 든 미국인들의 온라인 활동 비율은 낮지만, 55세 이상에서는 2008년에 1/3 가량이 웹을 활용하고, 온라인에서 보내는 시간은 오히려 젊은 층보다 많은 것으로 나타남

In [ ]:
age_gender[age_gender['age_bucket']  == '15-19']

## Language 탐색

In [ ]:
data13 = train.pivot_table("id", "language", aggfunc="count")

In [ ]:
data13['% of Total'] = (data13.id / data13.id.sum() * 100).round(4)
data13

In [ ]:
print((len(train[train['country_destination'] == 'US']) / len(train[train['language'] == 'en']) * 100))
print((len(train[train['country_destination'] == 'AU']) / len(train[train['language'] == 'en']) * 100))
print((len(train[train['country_destination'] == 'GB']) / len(train[train['language'] == 'en']) * 100))
print((len(train[train['country_destination'] == 'NDF']) / len(train[train['language'] == 'en']) * 100))

In [ ]:
Language 유의미하지 않을 수 있겠는데... 다들 영어여서...

In [ ]:
list_browser = train['first_browser'].groupby(train.first_browser).count()
df_browser = pd.DataFrame(list_browser)
br_list = df_browser[df_browser['first_browser']>100].index
temp = train['first_browser'].apply(lambda x : x if x in br_list else 'other_browser').replace('-unknown-', 'unknown')
first_browser = pd.get_dummies(temp)

In [ ]:
temp

## 4. date_first_booking 데이터 

In [ ]:
train.date_first_booking.isnull().sum()

In [ ]:
train[train.date_first_booking.notnull()].date_first_booking.describe()

In [ ]:
data4 = train.groupby('date_first_booking').aggregate({'id':'count'}).reset_index()
data4

## 5. Sign_up Data

In [ ]:
train.head()

In [ ]:
df_signup = train.loc[:, ['id', 'gender', 'age', 'language', 'signup_method', 'signup_flow', 'signup_app', 'country_destination']]
df_signup.tail()

In [ ]:
method_label = train['signup_method'].unique()
print(method_label)

In [ ]:
Image.open('sign_up.png')

In [ ]:
app_label = train['signup_app'].unique()
print(app_label)

In [ ]:
df_signup.tail()

In [ ]:
df_signup.pivot_table(columns='country_destination', index=['signup_method', 'signup_app'],values='id', aggfunc="count")

In [ ]:
len(df_signup)

In [ ]:
df_signup.isnull().sum()

In [ ]:
print(len(df_signup['age'].unique()))
xxx = df_signup.age.value_counts()
print(xxx[xxx == 26])

In [ ]:
df_signup2 = df_signup[df_signup.age < 101]
df_signup2.age.describe()

In [ ]:
data13 = df_signup2.pivot_table(columns='age', index=['signup_method', 'signup_app'],
                      values='id', aggfunc="count")
data13.fillna(0)

## 6. Affiliate (제휴 마케팅)

In [ ]:
df_affiliate = train.loc[:, ['id', 'gender', 'age', 'language', 'affiliate_channel', 'affiliate_provider', 'first_affiliate_tracked', 'country_destination']]
df_affiliate.tail()

In [ ]:
print(train['affiliate_channel'].unique())

seo : 검색최적화 / sem : 검색엔진마케팅 / content : 컨텐츠 제작광고

remarketing : 방문자 대상으로 광고

In [ ]:
print("제휴마케팅 제공자 : {}".format(len(train['affiliate_provider'].unique())))
print("=" * 80)
print(train['affiliate_provider'].unique())

In [ ]:
print("유입 경로 : {}".format(len(train['first_affiliate_tracked'].unique())))
print("=" * 80)
print(train['first_affiliate_tracked'].unique())

In [ ]:
df_affiliate.pivot_table(columns='country_destination', index=['affiliate_channel', 'affiliate_provider'],values='gender', aggfunc="count")

In [ ]:
df_affiliate2 = df_affiliate[df_affiliate.age < 101]
len(df_affiliate2.age) == 123116

In [ ]:
data14 = df_affiliate2.pivot_table(columns='age', index=['affiliate_channel', 'affiliate_provider'],
                      values='id', aggfunc="count")
data14.fillna(0)

In [ ]:
data14.to_csv('age_affiliate.csv', index_label=True, encoding='utf-8-sig')

## 7. First Access

In [ ]:
df_first = train.loc[:, ['gender', 'age', 'language', 'first_device_type', 'first_browser', 'country_destination']]
df_first.tail()

In [ ]:
print(train['first_device_type'].unique())

In [ ]:
print("browser 수 : {}".format(len(train['first_browser'].unique())))
print("=" * 80)
print(train['first_browser'].unique())

### PC 이용자

In [ ]:
first_device_type_pc = pd.concat([train[train.first_device_type == "Mac Desktop"], train[train.first_device_type == "Windows Desktop"], train[train.first_device_type == "Desktop (Other)"]])
first_device_type_pc.reset_index(drop=True, inplace=True)
first_device_type_pc.tail(3)

In [ ]:
data5 = first_device_type_pc.pivot_table(columns=['gender'], index=['first_device_type'], values='id', aggfunc="count")
data5

In [ ]:
plt.figure(figsize=(12,8))

data5.plot(kind='bar')
plt.xlabel('first_device_type')

plt.show()

In [ ]:
data6 = first_device_type_pc.pivot_table(columns=['first_device_type'], index=['country_destination'], values='id', aggfunc="count")
data6.plot(kind='bar')
plt.xlabel('country_destination')

plt.show()

### Tablet 이용자

In [ ]:
first_device_type_tab = pd.concat([train[train.first_device_type == "iPad"], train[train.first_device_type == "Android Tablet"]])
first_device_type_tab.reset_index(drop=True, inplace=True)
first_device_type_tab.tail(3)

In [ ]:
data7 = first_device_type_tab.pivot_table(columns=['gender'], index=['first_device_type'], values='id', aggfunc="count")
data7.plot(kind='bar')
plt.xlabel('gender')

plt.show()

In [ ]:
data8 = first_device_type_tab.pivot_table(columns=['first_device_type'], index=['country_destination'], values='id', aggfunc="count")
data8.plot(kind='bar')
plt.xlabel('Destination')

plt.show()

### ios 계열

In [ ]:
first_device_type_ios = pd.concat([train[train.first_device_type == "iPhone"], train[train.first_device_type == "iPad"]])
first_device_type_ios.reset_index(drop=True, inplace=True)
first_device_type_ios.tail(3)

In [ ]:
data9 = first_device_type_ios.pivot_table(columns=['gender'], index=['first_device_type'], values='id', aggfunc="count")
data9.plot(kind='bar')
plt.xlabel('gender')

plt.show()

In [ ]:
data10 = first_device_type_ios.pivot_table(columns=['first_device_type'], index=['country_destination'], values='id', aggfunc="count")
data10.plot(kind='bar')
plt.xlabel('Destination')

plt.show()

### Android 계열

In [ ]:
first_device_type_android = pd.concat([train[train.first_device_type == "Android Phone"], train[train.first_device_type == "Android Tablet"]])
first_device_type_android.reset_index(drop=True, inplace=True)
first_device_type_android.tail(3)

In [ ]:
data11 = first_device_type_android.pivot_table(columns=['gender'], index=['first_device_type'], values='id', aggfunc="count")
data11.plot(kind='bar')
plt.xlabel('gender')

plt.show()

In [ ]:
data12 = first_device_type_android.pivot_table(columns=['first_device_type'], index=['country_destination'], values='id', aggfunc="count")
data12.plot(kind='bar')
plt.xlabel('Destination')

plt.show()

## Countries Set

In [ ]:
countries.tail()

In [ ]:
train Country_destination에 맞춰서 countires의 distance_km,language_levenshtein_distance

거리는 여행경비와 연관이 있기때문에 충분히 중요한 변수가 될 수 있을 것 같다.

In [ ]:
import haversine

In [ ]:
countries['haversine_distance'] = countries.apply(lambda r: haversine.haversine(([36.966427, -95.844030]),
                                                                  (r['lat_destination'], r['lng_destination'])), axis=1)
countries

In [ ]:
distance_km : 미국을 기준으로 haversine_distance를 구한것
destination_km2 : 도착지 영토 크기

In [ ]:
Image.open('destination_km2.png')

In [ ]:
# import folium
# import folium.plugins as plugins
# from folium.plugins import MarkerCluster

In [ ]:
# center=[np.mean(countries['lat_destination']), np.mean(countries['lng_destination'])]

# m = folium.Map(
#     location = center,
#     zoom_start=0)


# for i in range(len(countries)):
#     location1=[countries.iloc[i, 1], countries.iloc[i, 2]]
#     folium.CircleMarker(
#         location=location1,
#         radius=1,
#         color = "blue"
#       ).add_to(m)
#     marker = folium.Marker(location=location1)
# m

# 나이 NaN값 채우는 방법론 idea.1

In [ ]:
df_chain = train.loc[:, ['age', 'first_device_type', 'first_browser', 'affiliate_channel', 'affiliate_provider', 'signup_method', 'signup_app']]
print(len(df_chain))
df_chain.tail()

In [ ]:
# NaN 포함
someone_like_you = pd.concat([df_chain[df_chain.first_device_type == "Mac Desktop"],
                              df_chain[df_chain.first_browser == "Safari"], 
                              df_chain[df_chain.affiliate_channel == "content"], 
                              df_chain[df_chain.affiliate_provider == "google"],
                              df_chain[df_chain.signup_method == "basic"], 
                              df_chain[df_chain.signup_app == "iOS"]])
someone_like_you.tail()

In [ ]:
someone_like_you = df_chain[df_chain.first_device_type == "Mac Desktop"]
someone_like_you = someone_like_you[someone_like_you.first_browser == "Safari"] 
someone_like_you = someone_like_you[someone_like_you.affiliate_channel == "content"] 
someone_like_you = someone_like_you[someone_like_you.affiliate_provider == "google"] 
someone_like_you = someone_like_you[someone_like_you.signup_method == "basic"] 
someone_like_you = someone_like_you[someone_like_you.signup_app == "Web"]

In [ ]:
someone_like_you = df_chain[(df_chain.first_device_type == "Mac Desktop") & (df_chain.first_browser == "Safari") ]

In [ ]:
len(someone_like_you)

In [ ]:
# NaN 미포함 & 100세 이상 제거
df_chain2 = df_chain[df_chain.age < 101]

In [ ]:
# NaN 미포함 & 100세 이상 제거
someone_like_you2 = df_chain2[df_chain2.first_device_type == "Mac Desktop"]
someone_like_you2 = someone_like_you2[someone_like_you2.first_browser == "Safari"] 
someone_like_you2 = someone_like_you2[someone_like_you2.affiliate_channel == "content"] 
someone_like_you2 = someone_like_you2[someone_like_you2.affiliate_provider == "google"] 
someone_like_you2 = someone_like_you2[someone_like_you2.signup_method == "basic"] 
someone_like_you2 = someone_like_you2[someone_like_you2.signup_app == "Web"]
someone_like_you2

In [ ]:
someone_like_you2.age.median(), someone_like_you2.age.mean()

In [ ]:
len(someone_like_you) - len(someone_like_you2)

In [ ]:
sessions.tail()

In [ ]:
print("Action 갯수 : {}".format(len(sessions['action'].unique())))
print("=" * 80)
print(sessions['action'].value_counts())

## 모델 준비단계

In [ ]:
# dfx = train.iloc[:, 0:15]
dfx = train.drop(columns=['country_destination'])
dfx

In [ ]:
target = train['country_destination']
target

In [ ]:
from sklearn.preprocessing import LabelEncoder

In [ ]:
le = LabelEncoder()
le.fit(target)
le.classes_

In [ ]:
y = le.transform(target)
y

In [ ]:
from sklearn.preprocessing import OneHotEncoder
ohe = OneHotEncoder()

In [ ]:
X

## 잡생각

여기어때
전체 회원 중 구매전환율(바로예약을 의미)은 40% 정도 된다.

In [ ]:
등록일 - 첫 부킹

#### Train Data Set에서 활용할 수 있는 요소로 생각할 수 있는 아이디어

1. 성별 / 연령 / 국적(거주지) 

2. 거주지와 목적지 간 이동시간 

3. 날짜정보 - 계절요인, 시기별

4. 시계열분석해서 시간에 따른 추세를 파악해볼까? (유행인 관광지)

인구ㆍ사회통계특성은 성별, 연령, 국적, 직업, 결혼, 소득, 학력, 거주지

여행특성은 동반유형, 동반자수, 여행형태, 방문유무, 방문횟수, 여행기간, 거주지와 목적지 간 이동시간

성격특성은 경험에 대한 개방성, 성실성, 외향성, 친화성, 신경증적 경향성을 포함한 세부정보 중 하나 이상을 포함하며, 

행동영향요인특성은 태도, 규범, 통제, 상황특성, 개인특성

In [ ]:
국토가 좁고 인구밀도가 높을 경우 국민들의 脫국내 심리가 커지고 국내여행 공급이 수요에 미치지 못하는 것
으로 추정(한국운행, 서비스 수지적자 지속원인과 대책(2008))

일본: 20대(27%) 미혼(42%)이 개별여행 방식(58%)으로 방문해 놀이/테마공원/온천 등 즐기기 (20%), 휴식(18%), 식도락(17%)을 즐기고자 하는 곳.

중국: 5-60대(57%) 남성(58%)이 단체 패키지(51%)로 가서 자연 풍경 감상(34%)하고, 역사/유적 감상(23%)을 주 목적으로 하는 곳.

유럽: 경제적으로 여유 있는(월수입 7백만원 이상 36%) 시니어들이 역사/유적(36%)과 도시 경관(22%)을 감상하기 위해 패키지여행(37%)을 계획 하는 곳.

미주(미국/하와이/캐나다): 고소득(월수입 7백만원 이상 41%) 남성(53%)이 친지/친구/친척 만나기(23%), 휴식(20%), 자연 풍경 감상(20%)등의 다양한 목적을 위해 개별여행(54%)을 계획하는 곳.

대양주: 3-40대(48%) 기혼(72%) 여성(54%)이 자연 풍경 감상(32%)과 휴식(28%)을 위해 찾는 곳. 상대적으로 에어텔 패키지가 많음(19%).

홍콩/마카오: 2-30대(56%) 미혼(43%) 여성(59%)이 도시 경관 감상(28%), 쇼핑(19%), 식도락(17%)을 위해 개별여행(52%)을 계획하는 곳.

사람들은 언어적으로 편한곳을 갈 것인가?
   : 다들 영어를 사용해서 구분이 힘듬.

성별에 따라 여행지 선택이 달라지는가?
   : 영향이 있는 것 같음
   
연령에 따라 여행지 선택이 달라지는가?

모바일 기기에 따른 경향성이 있는가?

첫 구매전환율 사람들은 어느 곳을 많이 가는가?


First-booking-date가 Nan 값인데...
Train에서 빼야할지??

## 전략

### 전략1 

1. Session ID를 Train / Test로 분리 (어느쪽에도 속하지 않는 ID가 있는가?)

2. Train에 NDF / 여행자 구분
   (테스트는 어차피 first booking date가 없다.)

3. NDF와 Destination을 session ID에 맞춰서 Merge

4. session에서 여행자의 Action/시간과 NDF의 Action/시간 분석

### 전략2

성별은 확실히 영향있는 것 같음

### 전략3 : 나이 NaN 값 채우기

Train data set에 있는 행동을 분석해서 나이 유추

### 전략4 : 어떻게 해서든지 출발지 찾기 (불가능할것 같지만)

age-gender의 인구와 countries의 영토면적을 이용해서 인구밀도 구함

인구밀도는 해외여행의 영향을 미친다고 학술, 국제기관 조사결과 나옴 

문제는 인구밀도를 구해도... 이사람이 어느나라 사람인지 알 수 있는 방법이 없다...

In [ ]:
계절, 시간 단위